### ISSUE TITLE
Should the selector in the method "get" from "HttpFoundation/ParameterBag" be changed?

### ISSUE THREAD

**Issue by enriquemorenotent** [symfony/symfony#14039]
I have been considering the `get` method in `HttpFoundation`. Basically we iterate around nested arrays using this kind of naming convention:

If we do `get("foo[bar][baz]")` we access `$this->params['foo']['bar']['baz']`.

The problem with this is that any string can be a key selector. Then how would you access a key that had the "[" or "]" chars as part of their key? For instance:

    $this->params["some[thing]"]

While we all will agree that this is a terrible name for a key, is it a possibility. Should we maybe reconsider the way `get` is called? I would propose the following:

    get(array("foo", "bar", "baz"));

What do you think?

**Comment by stof** [comment]
@enriquemorenotent By default, it does not perform any magic on the key, meaning that any key can be used. the magic only happens when changing the third argument explicitly

**Comment by enriquemorenotent** [comment]
But if the second or later argumetns have the characters `[` or `]`, you will not be able to reach it.

How would you do it to reach this?

    $this->params["foo"]["[bar]"]

**Comment by stof** [comment]
Well, if you want keys with ``[`` in them, don't rely on the magic of the deep get and get it explicitly instead (my code is missing error handling for missing keys):

```php
$this->get('foo')['[bar]'];
```

**Comment by enriquemorenotent** [comment]
Indeed is a way, but my point is that it is a confusing interface.

For instance, look at the following cases

    $this->get("[foo]");                 // Will return: $this->params["[foo]"];
    $this->get("[foo]", null, true);     // Will return: $this->params["foo"];

It is quite bizarre, and the rules aren't quite clear. Maybe we could make it better, but extending it. Handling the case when $path was an array. (Maybe eventually deprecating the other behavior?)

**Comment by enriquemorenotent** [comment]
How about implementing the interface `ArrayAccess`?

**Comment by stof** [comment]
Well, I actually never use the deep syntax. I don't like it (and it could probably have been a separate method instead of a flag in ``get``).

Regarding ArrayAccess, I'm not sure it is worth it. We cannot remove the ``get()`` method (it is a BC break), introducing a second way to achieve the retrieval (but without the possibility to specify a default value) only makes the API for confusing IMO.
And for PHP 5.4+, you can use ``$this->get('foo')['[bar]'];`` in case you are sure that keys are defined, which is not that defined from the ArrayAccess syntax (and the extra work needed for the error handling is the same with both syntaxes)

**Comment by enriquemorenotent** [comment]
I thought we could add a new way to access the values in the array, and eventually, through time, deprecate the old one.

I will be happy to do it, but first I want to make sure the change will be merged, so it is your call...

**Comment by fabpot** [comment]
I don't like this deep thing and I would really by very very happy if we were to deprecate this and remove it altogether from Symfony 3.0. One thing I'm sure about is that I don't want to invest time trying to make it "better". I fight against adding the same syntax for other methods (like remove) and I don't even remember why I accepted (or was it me?) the introduction of this symfony1-like syntax.

FYI: 89f60e0

**Comment by enriquemorenotent** [comment]
I would be happy to implement [ArrayAccess](http://php.net/manual/de/class.arrayaccess.php) in the class. I just do not know in which branch I should do the pull request.

**Comment by stof** [comment]
@enriquemorenotent I don't understand how ArrayAccess would change anything to the current syntax. It is **less powerful** than the current API, given it does not allow to specify a default value.

**Comment by stof** [comment]
FYI, the logic was originally added in https://github.com/symfony/symfony/pull/484 as a separate method before being moved to a flag in the get method in the commit you linked

**Comment by enriquemorenotent** [comment]
I just thought it would be nice if instead of setting the deep flag or the get method at all, we could just use `["key1"]["key2"]`.

The get method can co-live with that implementation, so we dont breack Backwards Compatibility.

**Comment by stof** [comment]
@enriquemorenotent ``->get('key1')['key2']`` gives you the same, without the need to provide a second API for the ParameterBag class

**Comment by enriquemorenotent** [comment]
Indeed, but it looks kinda "weird" and it feels a little unnatural, no? First a `()` and then a `[]`...

I thought we could do better

**Comment by jakzal** [comment]
:+1: for getting rid of the magic deep syntax. Since there's a language construct doing similar thing, it's not worth to maintain it on our side.

I also don't see much value in having ArrayAccess as I prefer explicit calls, and as mentioned by stof it would be less powerful (no default).

**Comment by enriquemorenotent** [comment]
Setting a default parameter is quite easy on an array

    $a = array("one" => 1);
    $b = $a["two"] ?: "default";
    echo $b;

will print

    default

**Comment by jakzal** [comment]
@enriquemorenotent this will emit a warning: *PHP Notice:  Undefined index: two in Command line code on line 2*

**Comment by enriquemorenotent** [comment]
Indeed. You are right.

I agree that a better method to set a default is needed,  but I still think that adding `ArrayAccess` would be a nice add on, since it is a nice native-looking syntax and some people might not even care about having a default.

**Comment by yguedidi** [comment]
@enriquemorenotent FYI, a native way to get a default value for an array item should be:
```php
// $_GET = array('one' => 1);
$params = $_GET + array('two' => 'default');
echo $params['two']; // default
```
http://3v4l.org/tRkMY

**Comment by fabpot** [comment]
Closing as the deep parameter has been deprecated and won't be available anymore in 3.0.

**Comment by adamelso** [comment]
I was trying to figure out why tests where failing for a library that relied on this feature where I attempted to provide support for Symfony 3. It took a while to find this issue, so I'm just referencing the related PR directly https://github.com/symfony/symfony/pull/16706

I have to say, this feature was very useful for apps that were served through the Facebook Page Tab - for some reason they enforced any query data to use the `app_data` parameter and ignored anything else, so nesting with `app_data[a]=1&app_data[b]=2` was [a suitable workaround.](https://github.com/archfizz/slotmachine/blob/master/README.markdown#assign-all-slots-to-one-http-query-parameter---working-with-apps-for-the-facebook-page-tab).

However, I haven't worked with Facebook apps for a long time and maybe this behaviour has changed since.
